distage/distage-framework-docker/src/main/scala/izumi/distage/docker/healthcheck/chain/PostgreSqlProtocolCheck.scala

```diff
@@ -0,0 +1,73 @@
+package izumi.distage.docker.healthcheck.chain
+
+import java.net.{InetSocketAddress, Socket}
+import java.nio.ByteBuffer
+
+import izumi.distage.docker.Docker.DockerPort
+import izumi.distage.docker.DockerContainer
+import izumi.distage.docker.healthcheck.ContainerHealthCheck
+import izumi.distage.docker.healthcheck.ContainerHealthCheck.HealthCheckResult
+import izumi.logstage.api.IzLogger
+import org.apache.commons.codec.binary.Hex
+
+final class PostgreSqlProtocolCheck[Tag](
+  userName: String = "postgres",
+  databaseName: String = "postgres",
+  port: DockerPort = DockerPort.TCP(5432),
+) extends ChainedContainerHealthCheck[Tag] {
+  override def check(
+    logger: IzLogger,
+    container: DockerContainer[Tag],
+    portStatus: HealthCheckResult.AvailableOnPorts,
+  ): ContainerHealthCheck.HealthCheckResult = {
+    portStatus.availablePorts.availablePorts.get(port) match {
+      case Some(value) if portStatus.requiredPortsAccessible =>
+        val startupMessage = genStartupMessage()
+        val socket = new Socket()
+        try {
+          val availablePort = value.head
+          socket.connect(new InetSocketAddress(availablePort.hostV4, availablePort.port), container.containerConfig.portProbeTimeout.toMillis.toInt)
+          logger.info(s"Checking PostgreSQL protocol on $port. Startup message: ${Hex.encodeHexString(startupMessage)}.")
+          val out = socket.getOutputStream
+          val in = socket.getInputStream
+          out.write(startupMessage)
+          val messageType = new String(in.readNBytes(1))
+          if (messageType == "R") {
+            logger.info(s"PostgreSQL protocol on $port is available.")
+            ContainerHealthCheck.HealthCheckResult.Available
+          } else {
+            logger.debug(s"PostgreSQL protocol on $port unavailable due to unknown message type: $messageType.")
+            ContainerHealthCheck.HealthCheckResult.Unavailable
+          }
+        } catch {
+          case t: Throwable =>
+            logger.debug(s"PostgreSQL protocol on $port unavailable due to unexpected exception. ${t.getMessage -> "Failure"}")
+            ContainerHealthCheck.HealthCheckResult.Unavailable
+        } finally {
+          socket.close()
+        }
+      case _ => ContainerHealthCheck.HealthCheckResult.Ignored
+    }
+  }
+
+  private def genStartupMessage(): Array[Byte] = {
+    def write(buffer: ByteBuffer, message: String): ByteBuffer = {
```

Comment:
That's cool. It would be a good idea to add a doc reference as a comment.

Reply:
Done. I've added small `toHex` function and place it into the `IzString`.